src/Microsoft.AspNetCore.ResponseCompression/ResponseCompressionProvider.cs

```diff
@@ -0,0 +1,102 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.ResponseCompression
+{
+    /// <inheritdoc />
+    public class ResponseCompressionProvider : IResponseCompressionProvider
+    {
+        private readonly ICompressionProvider[] _providers;
+        private readonly HashSet<string> _mimeTypes;
+
+        /// <summary>
+        /// If no compression providers are specified then GZip is used by default.
+        /// </summary>
+        /// <param name="providers">Compression providers to use, if any.</param>
+        /// <param name="options"></param>
+        public ResponseCompressionProvider(IEnumerable<ICompressionProvider> providers, IOptions<ResponseCompressionOptions> options)
+        {
+            _providers = providers.ToArray();
+            if (_providers.Length == 0)
+            {
+                _providers = new [] { new GzipCompressionProvider() };
+            }
+
+            if (options.Value.MimeTypes == null || !options.Value.MimeTypes.Any())
+            {
+                throw new InvalidOperationException("No mime types specified");
+            }
+            _mimeTypes = new HashSet<string>(options.Value.MimeTypes, StringComparer.OrdinalIgnoreCase);
+        }
+
+        /// <inheritdoc />
+        public virtual ICompressionProvider GetCompressionProvider(HttpContext context)
+        {
+            IList<StringWithQualityHeaderValue> unsorted;
+
+            // e.g. Accept-Encoding: gzip, deflate, sdch
+            var accept = context.Request.Headers[HeaderNames.AcceptEncoding];
+            if (!StringValues.IsNullOrEmpty(accept)
+                && StringWithQualityHeaderValue.TryParseList(accept, out unsorted)
+                && unsorted != null && unsorted.Count > 0)
+            {
+                // TODO PERF: clients don't usually include quality values so this sort will not have any effect. Fast-path?
+                var sorted = unsorted
+                    .Where(s => s.Quality.GetValueOrDefault(1) > 0)
+                    .OrderByDescending(s => s.Quality.GetValueOrDefault(1));
+
+                foreach (var encoding in sorted)
+                {
+                    if (string.Equals("*", encoding.Value, StringComparison.Ordinal))
+                    {
+                        return _providers[0];
+                    }
+                    if (string.Equals("identity", encoding.Value, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return null;
+                    }
+
+                    // There will rarely be more than three providers, and there's only one by default
+                    foreach (var provider in _providers)
```

Comment:
At least IE Edge or Chrome seem to fill AcceptEncoding with explicity values.
So I think it's probably faster to check `_providers` before "*" and "identity".